src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gcs_pub_sub.h"
+#include "ray/gcs/redis_context.h"
+
+namespace ray {
+namespace gcs {
+
+template <typename ID, typename Data>
+Status GcsPubSub<ID, Data>::Publish(const std::string &channel, const ID &id,
+                                    const Data &data, const StatusCallback &done) {
+  rpc::GcsMessage message;
```

Comment:
Is it possible to avoid using this `GcsMessage` to wrap id and data? For example, can we get the id from somewhere else, like the key?
I'm a bit concerned about too much memory copy here.